File: scala/src/main/scala/tddmicroexercises/textconvertor/HtmlPagesConverter.scala
package tddmicroexercises.textconvertor

import java.io.FileReader

import java.io.BufferedReader
import scala.util.control.Breaks._

class HtmlPagesConverter(val filename : String) {
  var breaks = List[Int](0)
  var reader = new BufferedReader(new FileReader(this.filename))
  var cumulativeCharCount = 0
  var line = reader.readLine()
  while (line != null)
  {
    cumulativeCharCount += line.length() + 1 // add one for the newline
    if (line.contains("PAGE_BREAK")) {
      val page_break_position = cumulativeCharCount
      breaks :+ page_break_position
    }
    line = reader.readLine()
  }
  reader.close()

  def getHtmlPage(page : Int) : String = {
    reader = new BufferedReader(new FileReader(filename))
    reader.skip(breaks(page))
    var htmlPage = new StringBuffer()
    var line = reader.readLine()
    while (line != null)
    {
      if (line.contains("PAGE_BREAK")) {
        break
      }
      htmlPage.append(StringEscapeUtils.escapeHtml(line))
      htmlPage.append("<br />")

      line = reader.readLine()
    }
    reader.close()
    return htmlPage.toString
  }
}

File: Cpp/TextConverter/source/HtmlPagesConverter.cpp

#include "../include/HtmlPagesConverter.h"
#include "../include/StringEscapeUtils.h"

#include <sstream>
#include <fstream>

HtmlPagesConverter::HtmlPagesConverter(std::string const& filename) : m_filename(filename)
{
    m_breaks.push_back(0);
    
    std::ifstream reader(m_filename);
    int cumulativeCharCount = 0;
    std::string line;
    while (std::getline(reader,line))
    {
        cumulativeCharCount += line.length() + 1; // add one for the newline
        if (line.find("PAGE_BREAK") != std::string::npos) 
        {
            int page_break_position = cumulativeCharCount;
            m_breaks.push_back(page_break_position);
        }
    }
    reader.close();
}

std::string HtmlPagesConverter::getHtmlPage(int page)
{
    std::ifstream reader(m_filename);
    reader.seekg(m_breaks[page]);
    std::ostringstream htmlPage;
    std::string line;
    while (std::getline(reader,line))
    {
        if (line.find("PAGE_BREAK") != std::string::npos) 
        {
            break;
        }
        htmlPage << StringEscapeUtils::escapeHtml(line);
        htmlPage << "<br />";
    }
    reader.close();
    return htmlPage.str();
}

std::string HtmlPagesConverter::getFilename() 
{
    return m_filename;
}
